File: src/main/java/org/voegtle/weatherstation/server/persistence/entities/AggregatedWeatherDataSet.java
package org.voegtle.weatherstation.server.persistence.entities;

import com.googlecode.objectify.annotation.Entity;
import com.googlecode.objectify.annotation.Id;
import com.googlecode.objectify.annotation.Ignore;
import com.googlecode.objectify.annotation.Index;
import org.voegtle.weatherstation.server.persistence.PeriodEnum;

import java.util.Date;

@Entity
public class AggregatedWeatherDataSet {
  @Id
  private Long id;

  @Index private Date date;

  @Index private PeriodEnum period;
  private boolean finished;

  private Date timeOfMinimum;
  private Date timeOfMaximum;

  private Float outsideTemperatureMin;
  private Float outsideTemperatureMax;
  private Float outsideTemperatureAverage;
  @Ignore
  private int outsideTemperatureCounter;

  private Float outsideHumidityMin;
  private Float outsideHumidityMax;
  private Float outsideHumidityAverage;
  @Ignore
  private int outsideHumidityCounter;

  private Float insideTemperatureMin;
  private Float insideTemperatureMax;
  private Float insideTemperatureAverage;
  @Ignore
  private int insideTemperatureCounter;

  private Float insideHumidityMin;
  private Float insideHumidityMax;
  private Float insideHumidityAverage;
  @Ignore
  private int insideHumidityCounter;

  private Float windspeedMin;
  private Float windspeedMax;
  private Float windspeedAverage;
  @Ignore
  private int windspeedCounter;

  private int rainCounter;
  private int rainDays;

  private Double kwh;

  public AggregatedWeatherDataSet() {
  }

  public AggregatedWeatherDataSet(Date date, PeriodEnum period) {
    this.date = date;
    this.period = period;
  }

  public void addOutsideTemperature(Float value, Date time) {
    if (value == null) {
      return;
    }
    if (outsideTemperatureAverage == null) {
      outsideTemperatureAverage = (float) 0.0;
    }

    outsideTemperatureCounter++;
    outsideTemperatureAverage += value;
    if (outsideTemperatureMin == null || outsideTemperatureMin > value) {
      outsideTemperatureMin = value;
      timeOfMinimum = time;
    }
    if (outsideTemperatureMax == null || outsideTemperatureMax < value) {
      outsideTemperatureMax = value;
      timeOfMaximum = time;
    }
  }

  public void addOutsideHumidity(Float value) {
    if (value == null) {
      return;
    }
    if (outsideHumidityAverage == null) {
      outsideHumidityAverage = (float) 0.0;
    }

    outsideHumidityCounter++;
    outsideHumidityAverage += value;
    if (outsideHumidityMin == null || outsideHumidityMin > value) {
      outsideHumidityMin = value;
    }
    if (outsideHumidityMax == null || outsideHumidityMax < value) {
      outsideHumidityMax = value;
    }
  }

  public void addInsideTemperature(Float value) {
    if (value == null) {
      return;
    }
    if (insideTemperatureAverage == null) {
      insideTemperatureAverage = (float) 0.0;
    }

    insideTemperatureCounter++;
    insideTemperatureAverage += value;
    if (insideTemperatureMin == null || insideTemperatureMin > value) {
      insideTemperatureMin = value;
    }
    if (insideTemperatureMax == null || insideTemperatureMax < value) {
      insideTemperatureMax = value;
    }
  }

  public void addInsideHumidity(Float value) {
    if (value == null) {
      return;
    }
    if (insideHumidityAverage == null) {
      insideHumidityAverage = (float) 0.0;
    }

    insideHumidityCounter++;
    insideHumidityAverage += value;
    if (insideHumidityMin == null || insideHumidityMin > value) {
      insideHumidityMin = value;
    }
    if (insideHumidityMax == null || insideHumidityMax < value) {
      insideHumidityMax = value;
    }
  }

  public void addWindspeed(Float value) {
    if (value == null) {
      return;
    }
    if (windspeedAverage == null) {
      windspeedAverage = (float) 0.0;
    }

    windspeedCounter++;
    windspeedAverage += value;
    if (windspeedMin == null || windspeedMin > value) {
      windspeedMin = value;
    }
    if (windspeedMax == null || windspeedMax < value) {
      windspeedMax = value;
    }
  }

  public void normalize() {
    if (outsideTemperatureCounter > 0) {
      outsideTemperatureAverage = outsideTemperatureAverage / outsideTemperatureCounter;
    }
    if (outsideHumidityCounter > 0) {
      outsideHumidityAverage = outsideHumidityAverage / outsideHumidityCounter;
    }

    if (insideTemperatureCounter > 0) {
      insideTemperatureAverage = insideTemperatureAverage / insideTemperatureCounter;
    }
    if (insideHumidityCounter > 0) {
      insideHumidityAverage = insideHumidityAverage / insideHumidityCounter;
    }
    if (windspeedCounter > 0) {
      windspeedAverage = windspeedAverage / windspeedCounter;
    }
  }

  public PeriodEnum getPeriod() {
    return period;
  }

  public void setPeriod(PeriodEnum period) {
    this.period = period;
  }

  public boolean isFinished() {
    return finished;
  }

  public void setFinished(boolean finished) {
    this.finished = finished;
  }

  public int getRainCounter() {
    return rainCounter;
  }

  public Float getRainAsFloat() {
    return rainCounter * 0.2f;
  }

  public void setRainCounter(int rainCounter) {
    this.rainCounter = rainCounter;
  }

  public int getRainDays() {
    return rainDays;
  }

  public void setRainDays(int rainDays) {
    this.rainDays = rainDays;
  }

  public Date getDate() {
    return date;
  }

  public Date getTimeOfMinimum() {
    return timeOfMinimum;
  }

  public Date getTimeOfMaximum() {
    return timeOfMaximum;
  }

  public Float getOutsideTemperatureMin() {
    return outsideTemperatureMin;
  }

  public Float getOutsideTemperatureMax() {
    return outsideTemperatureMax;
  }

  public Float getOutsideTemperatureAverage() {
    return outsideTemperatureAverage;
  }

  public Float getOutsideHumidityMin() {
    return outsideHumidityMin;
  }

  public Float getOutsideHumidityMax() {
    return outsideHumidityMax;
  }

  public Float getOutsideHumidityAverage() {
    return outsideHumidityAverage;
  }

  public Float getInsideTemperatureMin() {
    return insideTemperatureMin;
  }

  public Float getInsideTemperatureMax() {
    return insideTemperatureMax;
  }

  public Float getInsideTemperatureAverage() {
    return insideTemperatureAverage;
  }

  public Float getInsideHumidityMin() {
    return insideHumidityMin;
  }

  public Float getInsideHumidityMax() {
    return insideHumidityMax;
  }

  public Float getInsideHumidityAverage() {
    return insideHumidityAverage;
  }

  public Float getWindspeedMin() {
    return windspeedMin;
  }

  public void setWindspeedMin(Float windspeedMin) {
    this.windspeedMin = windspeedMin;
  }

  public Float getWindspeedMax() {
    return windspeedMax;
  }

  public void setWindspeedMax(Float windspeedMax) {
    this.windspeedMax = windspeedMax;
  }

  public Float getWindspeedAverage() {
    return windspeedAverage;
  }

  public void setWindspeedAverage(Float windspeedAverage) {
    this.windspeedAverage = windspeedAverage;
  }

  public Double getKwh() {
    return kwh;
  }

  public void setKwh(Double kwh) {
    this.kwh = kwh;
  }

  @Override
  public String toString() {
    return "AggregatedWeatherDataSet{" +
        "date=" + date +
        ", period=" + period +
        ", finished=" + finished +
        ", timeOfMinimum=" + timeOfMinimum +
        ", timeOfMaximum=" + timeOfMaximum +
        ", outsideTemperatureMin=" + outsideTemperatureMin +
        ", outsideTemperatureMax=" + outsideTemperatureMax +
        ", outsideTemperatureAverage=" + outsideTemperatureAverage +
        ", outsideTemperatureCounter=" + outsideTemperatureCounter +
        ", outsideHumidityMin=" + outsideHumidityMin +
        ", outsideHumidityMax=" + outsideHumidityMax +
        ", outsideHumidityAverage=" + outsideHumidityAverage +
        ", outsideHumidityCounter=" + outsideHumidityCounter +
        ", insideTemperatureMin=" + insideTemperatureMin +
        ", insideTemperatureMax=" + insideTemperatureMax +
        ", insideTemperatureAverage=" + insideTemperatureAverage +
        ", insideTemperatureCounter=" + insideTemperatureCounter +
        ", insideHumidityMin=" + insideHumidityMin +
        ", insideHumidityMax=" + insideHumidityMax +
        ", insideHumidityAverage=" + insideHumidityAverage +
        ", insideHumidityCounter=" + insideHumidityCounter +
        ", windspeedMin=" + windspeedMin +
        ", windspeedMax=" + windspeedMax +
        ", windspeedAverage=" + windspeedAverage +
        ", windspeedCounter=" + windspeedCounter +
        ", rainCounter=" + rainCounter +
        ", rainDays=" + rainDays +
        ", kwh=" + kwh +
        '}';
  }
}

File: src/main/java/org/voegtle/weatherstation/server/persistence/entities/AggregatedWeatherDataSet.kt
package org.voegtle.weatherstation.server.persistence.entities

import com.googlecode.objectify.annotation.Entity
import com.googlecode.objectify.annotation.Id
import com.googlecode.objectify.annotation.Ignore
import com.googlecode.objectify.annotation.Index
import java.util.*

@Entity
class AggregatedWeatherDataSet(@Id private var id: Long? = null,

                               @Index var date: Date = Date(),

                               var timeOfMinimum: Date? = null,
                               var timeOfMaximum: Date? = null,
                               var outsideTemperatureMin: Float? = null,
                               var outsideTemperatureMax: Float? = null,
                               var outsideTemperatureAverage: Float? = null,
                               @Ignore private var outsideTemperatureCounter: Int = 0,

                               var outsideHumidityMin: Float? = null,
                               var outsideHumidityMax: Float? = null,
                               var outsideHumidityAverage: Float? = null,
                               @Ignore private var outsideHumidityCounter: Int = 0,

                               var insideTemperatureMin: Float? = null,
                               var insideTemperatureMax: Float? = null,
                               var insideTemperatureAverage: Float? = null,
                               @Ignore private var insideTemperatureCounter: Int = 0,

                               var insideHumidityMin: Float? = null,
                               var insideHumidityMax: Float? = null,
                               var insideHumidityAverage: Float? = null,
                               @Ignore private var insideHumidityCounter: Int = 0,

                               var windspeedMin: Float? = null,
                               var windspeedMax: Float? = null,
                               var windspeedAverage: Float? = null,
                               @Ignore private var windspeedCounter: Int = 0,

                               var firstSolarRadiation: Date? = null,
                               var lastSolarRadiation: Date? = null,
                               var solarRadiationMax: Float? = null,
                               var solarRadiationMaxTime: Date? = null,
                               var solarRadiationTotal: Float? = null,

                               var totalPowerProduction: Float? = null,
                               var powerProductionMax: Float? = null,
                               var powerProductionMaxTime: Date? = null,

                               var dailyRain: Float? = null) {

  fun addOutsideTemperature(value: Float?, time: Date) {
    value?.let {
      outsideTemperatureCounter++
      outsideTemperatureAverage = (outsideTemperatureAverage ?: 0.0f) + it
      if (outsideTemperatureMin == null || outsideTemperatureMin!! > it) {
        outsideTemperatureMin = it
        timeOfMinimum = time
      }
      if (outsideTemperatureMax == null || outsideTemperatureMax!! < it) {
        outsideTemperatureMax = it
        timeOfMaximum = time
      }
    }
  }

  fun addOutsideHumidity(value: Float?) {
    value?.let {
      outsideHumidityCounter++
      outsideHumidityAverage = (outsideHumidityAverage ?: 0.0f) + it
      if (outsideHumidityMin == null || outsideHumidityMin!! > it) {
        outsideHumidityMin = it
      }
      if (outsideHumidityMax == null || outsideHumidityMax!! < it) {
        outsideHumidityMax = it
      }
    }
  }

  fun addInsideTemperature(value: Float?) {
    value?.let {
      insideTemperatureCounter++
      insideTemperatureAverage = (insideTemperatureAverage ?: 0.0f) + it
      if (insideTemperatureMin == null || insideTemperatureMin!! > it) {
        insideTemperatureMin = it
      }
      if (insideTemperatureMax == null || insideTemperatureMax!! < it) {
        insideTemperatureMax = it
      }
    }
  }

  fun addInsideHumidity(value: Float?) {
    value?.let {
      insideHumidityCounter++
      insideHumidityAverage = (insideHumidityAverage ?: 0.0f) + it
      if (insideHumidityMin == null || insideHumidityMin!! > it) {
        insideHumidityMin = it
      }
      if (insideHumidityMax == null || insideHumidityMax!! < it) {
        insideHumidityMax = it
      }
    }
  }

  fun addWindspeed(value: Float?) {
    value?.let {
      windspeedCounter++
      windspeedAverage = (windspeedAverage ?: 0.0f) + it
      if (windspeedMin == null || windspeedMin!! > it) {
        windspeedMin = it
      }
    }
  }

  fun updateWindspeedMax(value: Float?) {
    value?.let {
      if (windspeedMax == null || windspeedMax!! < it) {
        windspeedMax = it
      }
    }
  }

  fun normalize() {
    if (outsideTemperatureCounter > 0) {
      outsideTemperatureAverage = outsideTemperatureAverage!! / outsideTemperatureCounter
    }
    if (outsideHumidityCounter > 0) {
      outsideHumidityAverage = outsideHumidityAverage!! / outsideHumidityCounter
    }

    if (insideTemperatureCounter > 0) {
      insideTemperatureAverage = insideTemperatureAverage!! / insideTemperatureCounter
    }
    if (insideHumidityCounter > 0) {
      insideHumidityAverage = insideHumidityAverage!! / insideHumidityCounter
    }
    if (windspeedCounter > 0) {
      windspeedAverage = windspeedAverage!! / windspeedCounter
    }
  }

  fun addSolarRadiation(solarRadiation: Float?) {
    solarRadiation?.let {
      solarRadiationTotal = (solarRadiationTotal ?: 0.0f) + (it / 4) // solarRadiation = durchschnittliche Sonneneinstrahlung
                                                                     // der letzten viertel Stunde
    }
  }

  fun updateSolarRadiationMax(solarRadiation: Float?, timestamp: Date) {
    solarRadiation?.let {
      if (solarRadiationMax == null || solarRadiationMax!! < it) {
        solarRadiationMax = it
        solarRadiationMaxTime = timestamp
      }
      if (it > 0) {
        firstSolarRadiation = firstSolarRadiation ?: timestamp
        lastSolarRadiation = timestamp
      }
    }
  }

  fun updatePowerProductionMax(powerProduction: Float?, timestamp: Date) {
    powerProduction?.let {
      if (powerProductionMax == null || powerProductionMax!! < it) {
        powerProductionMax = powerProduction
        powerProductionMaxTime = timestamp
      }
    }
  }
}
